contents/blog/temporal-exports.md

```diff
@@ -0,0 +1,118 @@
+---
+title: Why and how we moved exports to Temporal
+date: 2023-08-04
+author: ["tomas-farias", "ian-vanagas"]
+showTitle: true
+rootpage: /blog
+sidebar: Blog
+hideAnchor: true
+featuredImage: ../images/blog/posthog-engineering-blog.png
+featuredImageType: full
+category: Engineering
+---
+
+Letting users maintain ownership of their data is a key value of PostHog. Although we aspire to become the single source of truth for all your data needs, we realize we don’t fulfill everything everyone wants to do (yet). Also, we know that even if we were the single source of truth, users  still want to use their data on other platforms.
+
+Our export system provides the ability to maintain ownership of all of your data in PostHog and move it where you like. Unfortunately, it didn’t scale as well as we wanted. Luckily for you, we recently launched our [newly upgraded exports system](/docs/cdp/batch-exports), built using Temporal. This post covers why and how we made the change to Temporal.
+
+## The trouble with exports
+
+Although exports worked well most of the time, they were causing trouble for some customers. They were failing, erroring, and dropping data. Fixing these issues was taking up a significant amount of the pipeline team’s time. 
+
+Although stateless plugins (like GeoIP) worked well with our [plugin service structure](/blog/how-we-built-an-app-server), **stateful** plugins (like exports) didn’t. This is because they depended on external resources that can degrade or fail. Dealing with this required timeout and retry logic that was difficult  to build into the existing structure.
+
+Our old export system scaled as far as possible. It worked well for smaller users, but as the export requirements grew, it started to show issues. Chiefly among them included:
+
+### Unreliable (not retrying)
+
+The biggest issue was not handling errors gracefully, which made the entire system unreliable. This is largely due to the statefulness of exports. A failure meant they lost their state which was unrecoverable.
```

Comment:
I don't think this is the case: Export apps do maintain a state that is not lost on failure. This state was stored in the `PluginStorage` model and contained information about how much progress had the export been able to achieve.

The biggest issue in my experience wasn't that exports would fail but that the export manager had no way of reliably detecting whether an export had failed, so it would eagerly issue retries. This would cause multiple copies of the same export job to be running at the same time. The problem was that each copy used **the same** state (i.e. a row in the `PluginStorage` table) and would race against each other to update it. Thus, whenever a new copy started it would reset the progress to zero, and other running copies would then resume from the beginning again.

Reply:
What I just wrote applies to historical exports.

Reply:
Ok, rewrote to focus on failing and re-running race condition.